Interpolating string dangers (#185)
andycroll/andycroll.com: _posts/2023/2023-05-29-really-really-really-do-not-interpolate-strings-into-active-record-methods.md

---
title: "Really, Really, Really Don’t Interpolate Strings into Active Record Methods"
description: "It’s risky to expose your database to SQL Injections"
layout: article
category: ruby
image:
  base: '2023/really-really-really-do-not-interpolate-strings-into-active-record-methods'
  alt: "Tables in a courtyard"
  credit: "Dimitra Peppa"
  source: "https://unsplash.com/photos/-abBaVOMsBk"

---

> Protecting yourself, and your application, against malicious users is a key responsibility as a developer. The built-in security provided by a well-maintained framework like Rails is a great reason to use one.

Reviewer: ```suggestion
Protecting your application against malicious users is one of your key responsibilites as a developer. The built-in security provided by a well-maintained framework, such as Rails, is an excellent reason to use one.
```

> This is particularly true of the protection afforded within Active Record for sanitizing user input before it is written to your database. However there are ways to pass strings directly to Active Record scopes, when you need to, however that power should be used _very_ carefully.

Reviewer: ```suggestion
This is particularly true of the protection afforded within Active Record for sanitizing user input before it is written to your database. However there are ways to pass strings directly to Active Record scopes when you need to, but that power should be used _very_ sparingly and carefully.
```

## Instead of…

> …using strings in your arguments to Active Record:

Reviewer: ```suggestion
…using strings in any arguments sent to Active Record:
```

```ruby
User.delete_by("id = #{params[:id]}")

> User.where("email = #{params[:email]}")

Reviewer: Would people really opt to do this instead of your suggested alternative, which to me is the no-brainer default??

Author: I've seen similar. In old codebases.

```

## Use…

…hash-based variants of the same methods:

```ruby
User.delete_by(id: params[:id])
User.where(email: params[:email])
```

## Why?

> Rails is a [sharp knife](https://rubyonrails.org/doctrine#provide-sharp-knives), while it does a lot for developers it also allows you the flexibility to bend the framework to your use case. In this case passing strings to Active Record methods.

Reviewer: ```suggestion
Rails is a [sharp knife](https://rubyonrails.org/doctrine#provide-sharp-knives). While it does a lot for developers, it also allows you the flexibility to bend the framework to your use case. In this case: passing strings to Active Record methods.
```

Using strings with interpolated (and user-provided) parameters opens you up to SQL injection attacks.

```ruby
# user-provided parameter
params[:id] = "1) OR 1=1--"
User.delete_by("id = #{params[:id]}")
#=> User Delete All (4.2ms)  DELETE FROM "users" WHERE (id = 1) OR 1=1--)
```

> The `1=1` part of the user-provided string is _always_ true and would run an SQL command to drop every user in your database. Not good.

Reviewer: ```suggestion
The `1=1` part of the user-provided string above is _always_ true and so would trigger an SQL command that drops every user in your database. Not good.
```

> Interpolating values directly into the arguments can lead to unpredictable behavior and results, not just malicious destructive example like above. You might leak information you hadn't intended.

Reviewer: ```suggestion
Interpolating values directly into the arguments can lead to unpredictable behaviour and results, not just malicious destructive examples like the one above. For example, you might leak information you hadn't intended to.
```

```ruby
params[:email] == 

> User.where("email = #{params[:email]}")

Reviewer: Is this an example of the above leaking? Not sure from this how it leaks.

Author: Fixed.

```

> Lastly, the string-based arguments make your code harder to read and understand, the syntax for the hash-based syntax is much easier to understand.

Reviewer: ```suggestion
Finally, the string-based arguments make your code harder to read and understand. The syntax for the hash-based approach is much easier to understand.
```

## Why not?

These are contrived examples. And yet real weaknesses.

> If you’re unable to use the hash-style for the specific query you arequire and you really, really, _really,_ know what you’re doing, then use the string-based arguments. But be careful.

Reviewer: ```suggestion
If you’re unable to use the hash-style for the specific query you require and you really, really, _really,_ know what you’re doing, then use the string-based arguments. But be careful.
```

Reviewer: Any examples of when you might really need to do this? Even if just a text-based explanation of them.

I also wonder whether it's worth leading with: these are super contrived examples to demo the point as at times people do reach for this technique...

I just think this whole piece would be more powerful if you used an example where people might think interpolating is the best option but then you show: oh look, here's how to do it the hash-based way (the way that we all know is the default).

```

## Why?

Rails is a [sharp knife](https://rubyonrails.org/doctrine#provide-sharp-knives). While it does a lot for developers, it also allows you the flexibility to bend the framework to your use case. In this case: passing strings to Active Record methods.

You’re unlikely to reach for string interpolation in the specific examples above given the breadth of support for straightforward database actions in Active Record, but when you have a long and complex SQL query, it is easy to forget to sanitize any user input.

Using strings with interpolated (and user-provided) parameters opens you up to SQL injection attacks.

```ruby
# user-provided parameter
params[:id] = "1) OR 1=1--"
User.delete_by("id = #{params[:id]}")
#=> User Delete All (4.2ms)  DELETE FROM "users" WHERE (id = 1) OR 1=1--)
```

The `1=1` part of the user-provided string above is _always_ true and so would trigger an SQL command that drops every user in your database. Not good.

Interpolating values directly into the arguments can lead to unpredictable behaviour and results, not just malicious destructive examples like the one above. For example, you might leak information you hadn't intended to.

```ruby

> params[:q] = "'' OR '1'='1';"

Reviewer: Wait, what is this example meant to be showing?

Author: Explained in https://github.com/andycroll/andycroll.com/pull/185/commits/de01ecf64c5bb6e8523357d5c19d829456a841cb

User.where("email = #{params[:q]}")
```

> Finally, the string-based arguments make your code harder to read and understand. The syntax for the hash-based approach is much easier to understand.

Reviewer: Maybe "much more accessible" in order to avoid repeating "understand"?

Author: ```suggestion
Finally, the string-based arguments make your code harder to read and understand. The syntax for the hash-based approach is much easier to comprehend.
```
